File: services/api-ts/src/adapters/http.ts
import type { IncomingMessage, ServerResponse } from "node:http";
import type { TaskService } from "../application/task-service.ts";
import { DomainError, parseStatus } from "../domain/task.ts";

/** A request body larger than this is refused before it is parsed. */
const MAX_BODY_BYTES = 1024 * 1024;

const STATUS_BY_CODE = {
  EMPTY_TITLE: 422,
  TITLE_TOO_LONG: 422,
  UNKNOWN_STATUS: 422,
  INVALID_TRANSITION: 409,
  NOT_FOUND: 404,
} as const satisfies Record<DomainError["code"], number>;

class BadRequest extends Error {}

export type Log = (entry: Record<string, unknown>) => void;

/**
 * The HTTP transport: decode the request, call a use case, map the outcome to a response. It holds
 * no business rules, and the API it serves matches api-go's route for route.
 */
export function createHandler(service: TaskService, log: Log) {
  return async (req: IncomingMessage, res: ServerResponse): Promise<void> => {
    try {
      await route(service, req, res);
    } catch (err) {
      if (err instanceof DomainError) {
        send(res, STATUS_BY_CODE[err.code], { error: err.message });
      } else if (err instanceof BadRequest) {
        send(res, 400, { error: err.message });
      } else {
        // The detail is logged and never returned to the client.
        log({ level: "error", msg: "request failed", method: req.method, path: req.url, error: String(err) });
        send(res, 500, { error: "internal error" });
      }
    }
  };
}

async function route(service: TaskService, req: IncomingMessage, res: ServerResponse): Promise<void> {
  // HEAD is answered wherever GET is, as HTTP requires; Node leaves the body off the response itself.
  const method = req.method === "HEAD" ? "GET" : (req.method ?? "GET");
  const path = new URL(req.url ?? "/", "http://localhost").pathname;

  if (path === "/healthz") {
    return method === "GET" ? send(res, 200, { status: "ok" }) : notAllowed(res);
  }
  if (path === "/api/tasks") {
    if (method === "GET") return send(res, 200, await service.list());
    if (method === "POST") {
      const body = await readObject(req, ["title"]);
      return send(res, 201, await service.create(optionalString(body, "title")));
    }
    return notAllowed(res);
  }

  const match = /^\/api\/tasks\/([^/]+)(\/status)?$/.exec(path);
  if (match?.[1] === undefined) return send(res, 404, { error: "not found" });
  const id = decodeSegment(match[1]);
  if (match[2] === undefined) {
    return method === "GET" ? send(res, 200, await service.get(id)) : notAllowed(res);
  }
  if (method !== "PATCH") return notAllowed(res);
  const body = await readObject(req, ["status"]);
  return send(res, 200, await service.transition(id, parseStatus(body["status"])));
}

/** Reads a size-bounded JSON object and refuses fields the endpoint does not document. */
async function readObject(req: IncomingMessage, allowed: readonly string[]): Promise<Record<string, unknown>> {
  const chunks: Buffer[] = [];
  let size = 0;
  for await (const chunk of req) {
    size += (chunk as Buffer).length;
    // Keep draining past the limit instead of destroying the stream, so the client still receives
    // the 400; the server's requestTimeout bounds how long that can take.
    if (size <= MAX_BODY_BYTES) chunks.push(chunk as Buffer);
  }
  const invalid = new BadRequest("request body must be a JSON object with only the documented fields");
  if (size > MAX_BODY_BYTES) throw invalid;
  let body: unknown;
  try {
    body = JSON.parse(Buffer.concat(chunks).toString("utf8"));
  } catch {
    throw invalid;
  }
  if (typeof body !== "object" || body === null || Array.isArray(body)) throw invalid;
  if (Object.keys(body).some((key) => !allowed.includes(key))) throw invalid;
  return body as Record<string, unknown>;
}

/** A missing or null field reads as empty, which the domain rejects; a field of another type is malformed. */
function optionalString(body: Record<string, unknown>, key: string): string {
  const value = body[key];
  if (value === undefined || value === null) return "";
  if (typeof value !== "string") throw new BadRequest(`${key} must be a string`);
  return value;
}

function decodeSegment(segment: string): string {
  try {
    return decodeURIComponent(segment);
  } catch {
    throw new BadRequest("malformed path");
  }
}

function notAllowed(res: ServerResponse): void {
  send(res, 405, { error: "method not allowed" });
}

function send(res: ServerResponse, status: number, body: unknown): void {
  res.writeHead(status, { "content-type": "application/json" }).end(JSON.stringify(body));
}

File: services/api-py/src/api_py/adapters/http.py
"""The HTTP transport: decode the request, call a use case, map the outcome to a response.

It holds no business rules, and the API it serves matches api-go and api-ts route for route. It is a
plain WSGI application, so the same object runs under the development server in main.py and under a
production server such as gunicorn or waitress with no code change.
"""

from __future__ import annotations

import json
import re
from collections.abc import Callable, Iterable
from typing import Any, Final
from urllib.parse import unquote

from api_py.application.task_service import TaskService
from api_py.domain.task import DomainError, DomainErrorCode, Task, parse_status

# A request body larger than this is refused before it is parsed.
MAX_BODY_BYTES: Final = 1024 * 1024
# How much of a refused body is still read, so the client can receive the 400 instead of a reset
# connection. The standard-library server has no request timeout, so this bounds the work instead.
MAX_DRAIN_BYTES: Final = 8 * MAX_BODY_BYTES

STATUS_BY_CODE: Final[dict[DomainErrorCode, int]] = {
    "EMPTY_TITLE": 422,
    "TITLE_TOO_LONG": 422,
    "UNKNOWN_STATUS": 422,
    "INVALID_TRANSITION": 409,
    "NOT_FOUND": 404,
}

REASON: Final[dict[int, str]] = {
    200: "OK",
    201: "Created",
    400: "Bad Request",
    404: "Not Found",
    405: "Method Not Allowed",
    409: "Conflict",
    422: "Unprocessable Content",
    500: "Internal Server Error",
}

_TASK_PATH = re.compile(r"^/api/tasks/([^/]+)(/status)?$")
_MALFORMED_ESCAPE = re.compile(r"%(?![0-9A-Fa-f]{2})")

StartResponse = Callable[[str, list[tuple[str, str]]], Any]
WSGIApplication = Callable[[dict[str, Any], StartResponse], Iterable[bytes]]
Log = Callable[[dict[str, object]], None]


class BadRequestError(Exception):
    """A request the transport cannot read. The message is written for the client."""

    def __init__(self, message: str) -> None:
        super().__init__(message)
        self.message: Final = message


def as_json(task: Task) -> dict[str, str]:
    """The wire shape, which is camelCase in every one of the three services."""
    return {
        "id": task.id,
        "title": task.title,
        "status": task.status,
        "createdAt": task.created_at,
        "updatedAt": task.updated_at,
    }


def create_app(service: TaskService, log: Log) -> WSGIApplication:
    def app(environ: dict[str, Any], start_response: StartResponse) -> Iterable[bytes]:
        try:
            status, body = _route(service, environ)
        except DomainError as err:
            status, body = STATUS_BY_CODE[err.code], {"error": err.message}
        except BadRequestError as err:
            status, body = 400, {"error": err.message}
        except Exception as err:  # noqa: BLE001 - the boundary: nothing escapes as a stack trace
            # The detail is logged and never returned to the client.
            log(
                {
                    "level": "error",
                    "msg": "request failed",
                    "method": environ.get("REQUEST_METHOD"),
                    "path": environ.get("PATH_INFO"),
                    "error": str(err),
                }
            )
            status, body = 500, {"error": "internal error"}

        payload = json.dumps(body).encode()
        start_response(
            f"{status} {REASON[status]}",
            [("content-type", "application/json"), ("content-length", str(len(payload)))],
        )
        # A HEAD response describes the GET response and carries no body; WSGI servers send whatever
        # the application returns, so leaving it out is this code's job.
        return [b""] if environ.get("REQUEST_METHOD") == "HEAD" else [payload]

    return app


def _route(service: TaskService, environ: dict[str, Any]) -> tuple[int, object]:
    # HEAD is answered wherever GET is, as HTTP requires.
    method = environ.get("REQUEST_METHOD", "GET")
    method = "GET" if method == "HEAD" else method
    path = _path(environ)

    if path == "/healthz":
        return (200, {"status": "ok"}) if method == "GET" else _not_allowed()

    if path == "/api/tasks":
        if method == "GET":
            return 200, [as_json(task) for task in service.list()]
        if method == "POST":
            body = _read_object(environ, ["title"])
            return 201, as_json(service.create(_optional_string(body, "title")))
        return _not_allowed()

    match = _TASK_PATH.match(path)
    if match is None:
        return 404, {"error": "not found"}

    if _MALFORMED_ESCAPE.search(match.group(1)):
        raise BadRequestError("malformed path")
    task_id = unquote(match.group(1))
    if match.group(2) is None:
        return (200, as_json(service.get(task_id))) if method == "GET" else _not_allowed()

    if method != "PATCH":
        return _not_allowed()
    body = _read_object(environ, ["status"])
    return 200, as_json(service.transition(task_id, parse_status(body.get("status"))))


def _path(environ: dict[str, Any]) -> str:
    """The request path, still percent-encoded where the server exposes it.

    WSGI decodes PATH_INFO before routing, which would turn an encoded slash inside an id into a path
    separator. Servers that keep the raw target (gunicorn's RAW_URI, some others' REQUEST_URI) let the
    route match first and the segment be decoded after, as api-go and api-ts do.
    """
    raw = environ.get("RAW_URI") or environ.get("REQUEST_URI")
    if isinstance(raw, str) and raw != "":
        return raw.split("?", 1)[0]
    path = environ.get("PATH_INFO", "/")
    return path if isinstance(path, str) and path != "" else "/"


def _not_allowed() -> tuple[int, object]:
    return 405, {"error": "method not allowed"}


def _read_object(environ: dict[str, Any], allowed: list[str]) -> dict[str, object]:
    """Reads a size-bounded JSON object and refuses fields the endpoint does not document."""
    invalid = BadRequestError("request body must be a JSON object with only the documented fields")
    try:
        length = int(environ.get("CONTENT_LENGTH") or 0)
    except ValueError:
        raise invalid from None
    stream = environ.get("wsgi.input")
    if length > MAX_BODY_BYTES:
        _drain(stream, min(length, MAX_DRAIN_BYTES))
        raise invalid
    raw = stream.read(length) if stream is not None and length > 0 else b""
    # An empty body is not an empty object: api-go and api-ts refuse it as malformed, and so does this.
    if raw.strip() == b"":
        raise invalid
    try:
        body = json.loads(raw)
    except ValueError:
        raise invalid from None
    if not isinstance(body, dict):
        raise invalid
    if any(key not in allowed for key in body):
        raise invalid
    return body


def _drain(stream: Any, remaining: int) -> None:
    """Reads and discards what the client is still sending, in bounded chunks."""
    while stream is not None and remaining > 0:
        chunk = stream.read(min(remaining, 64 * 1024))
        if not chunk:
            return
        remaining -= len(chunk)


def _optional_string(body: dict[str, object], key: str) -> str:
    """A missing field reads as empty, which the domain rejects; a wrong type is malformed."""
    value = body.get(key)
    if value is None:
        return ""
    if not isinstance(value, str):
        raise BadRequestError(f"{key} must be a string")
    return value
